llvmPackages_{12,13,14,15,16,17}: drop

emilazy opened the pull request:
We currently package ten stable versions of LLVM going back 4½ years. Older versions of LLVM don’t get upstream fixes, and require significant derivation complexity and patching to keep working, especially the further back we go.  Dropping them eases the burden of LLVM maintenance and saves a bunch of expensive compiler builds on Hydra.

The packages marked broken or dropped here will be the only remaining users in the tree after the remaining in‐flight PRs are merged. I made varying degrees of effort to get all of them building with newer versions, as per the list of preceding PRs below, but was not successful in these cases. The upstreams have varying degrees of activity. If anyone can get them building with LLVM ≥ 18 then they can be removed from this PR or re‐added after it is merged. I’ve left some notes in the commit messages and code comments where applicable for anyone who wants to try. I do hope that we will drop old compilers more regularly going forward, though, so I wouldn’t recommend making a one‐off heroic effort if it’s expected that it will end up bitrotting again a couple releases down the line. In particular, I hope that we will be able to drop at least LLVM 18 for 26.05.

This concludes my work to drop old compilers for 25.11 – reviews of open PRs are appreciated:

* <https://github.com/NixOS/nixpkgs/pull/434139>
* <https://github.com/NixOS/nixpkgs/pull/434553>
* <https://github.com/NixOS/nixpkgs/pull/434642>
* <https://github.com/NixOS/nixpkgs/pull/434644>
* <https://github.com/NixOS/nixpkgs/pull/434649>
* <https://github.com/NixOS/nixpkgs/pull/434776>
* <https://github.com/NixOS/nixpkgs/pull/434788>
* <https://github.com/NixOS/nixpkgs/pull/434827>
* <https://github.com/NixOS/nixpkgs/pull/435019>
* <https://github.com/NixOS/nixpkgs/pull/435063>
* <https://github.com/NixOS/nixpkgs/pull/435096>
* <https://github.com/NixOS/nixpkgs/pull/435110>
* <https://github.com/NixOS/nixpkgs/pull/435148>
* <https://github.com/NixOS/nixpkgs/pull/435927>
* <https://github.com/NixOS/nixpkgs/pull/435932>
* <https://github.com/NixOS/nixpkgs/pull/437152>
* <https://github.com/NixOS/nixpkgs/pull/437364>
* <https://github.com/NixOS/nixpkgs/pull/439408>
* <https://github.com/NixOS/nixpkgs/pull/439740>
* <https://github.com/NixOS/nixpkgs/pull/439928>
* <https://github.com/NixOS/nixpkgs/pull/440271>
* <https://github.com/NixOS/nixpkgs/pull/440272>

cc’ing maintainers of packages marked broken or dropped in this PR:

* `shader-slang` (LLVM backend only) – @niklaskorz
* `fpc` (`x86_64-darwin` only) – @7c6f434c
* `mrustc-bootstrap` – @progval @r-burns
* `spirv-llvm-translator` (LLVM 14–17 versions only) – @gloaming
* `bfc` – @figsoda
* `ikos` – @AtnNn
* `oclgrind` – @athas
* `cone` – nobody 😢
* `dale` – @amiloradovsky
* `wavm` – @ereslibre
* `qrscan` – @sayanarijit
* `hobbes` – @kthielen @thmzlt

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy committed:
llvmPackages_16: drop

emilazy committed:
fpc: mark broken with Clang on x86‐64

This is only really relevant to `x86_64-darwin`, and it’s not worth
keeping an unmaintained LLVM around for given the limited remaining
lifespan of the platform.

emilazy committed:
llvmPackages_17: drop

emilazy committed:
bfc: drop

emilazy committed:
shader-slang: mark LLVM support broken

emilazy committed:
ikos: drop

emilazy committed:
cone: drop

emilazy committed:
oclgrind: drop

emilazy committed:
llvmPackages_13: drop

emilazy committed:
llvmPackages_14: drop

emilazy committed:
dale: drop

Open issue from 2023 about supporting LLVM 15 with no progress,
no activity in the repository in almost a year.

emilazy committed:
llvmPackages_12: drop

emilazy committed:
wavm: drop

emilazy committed:
qrscan: drop

emilazy committed:
llvmPackages_15: drop

emilazy committed:
hobbes: drop

emilazy committed:
mrustc-bootstrap: mark broken

emilazy committed:
spirv-llvm-translator: drop versions for LLVM 14–17

emilazy committed:
haskell.compiler.ghc{948,963,967,984,9101,9102,9121,9122}: bump LLVM

Bump the LLVM version used by our GHC packages in preparation for
the LLVM drops. This breaks the LLVM backend; it will be fixed soon
by backporting patches, but that will require a mass rebuild.

emilazy committed:
llvmPackages: remove redundant comment

emilazy committed:
llvmPackages: remove obsolete version conditionals

emilazy committed:
llvmPackages: add comment for mistake

emilazy committed:
llvmPackages: remove unreferenced patches

niklaskorz reviewed (approved):
re shader-slang, I'll reference the discussion in https://github.com/NixOS/nixpkgs/pull/436961

TL;DR: the LLVM mode is in a weird (non-functional) state anyway, so we'll probably have to refactor or remove it. Upstream currently doesn't seem capable to keep up with recent LLVM versions, so removing `withLLVM` might even be preferred for now.

I'm fine with simply marking it broken here so we can defer that decision.

emilazy commented:
Cool, thank you for the pointer! That is one of the packages I was most worried about so it is good to hear it at least won't be getting any more broken. I think I did not spend too long seeing how much work it would be to get it building with newer LLVM versions, so it's possible that it might be an upstreamable patch away from working. But I lack the context to know whether it'd be worth the energy, beyond what they have written down in [their docs](https://github.com/shader-slang/slang/blob/5500f11768d4d93eef4dfcecf0821fee747bf1a4/docs/cpu-target.md#slang-llvm). I suppose if it's like most use cases for shaders then CPU execution isn't a priority.

philiptaron reviewed (approved):
Thanks, @emilazy.

alyssais commented on pkgs/development/compilers/llvm/common/compiler-rt/default.nix:
```diff
@@ -72,64 +72,54 @@ stdenv.mkDerivation (finalAttrs: {
+    lib.optional (lib.versionOlder release_version "18")
+      # Prevent a compilation error on darwin
+      (getVersionFile "compiler-rt/darwin-targetconditionals.patch")
+  ++ [
```
Here too — in fact it can even be fixed in this PR, after the 17 removal.

alyssais commented on pkgs/development/compilers/llvm/common/compiler-rt/default.nix:
```diff
@@ -41,10 +41,7 @@ let
-    stdenv.hostPlatform.isDarwin
-    && stdenv.hostPlatform.isStatic
-    && lib.versionAtLeast release_version "16";
+  isDarwinStatic = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isStatic;
```
Maybe we should just inline this now?

alyssais commented on pkgs/development/compilers/llvm/common/libcxx/default.nix:
```diff
@@ -163,37 +160,14 @@ stdenv.mkDerivation (
-            "/llvm/cmake/"
-          ];
-        })
-      ]
```
This removal doesn't look right in the commit it's in.

emilazy replied:
You mean, because it leaves the Musl patch when it shouldn’t, and I incorrectly deleted the Musl patch as unreferenced in an earlier commit? Yes, good catch; will fix.

emilazy replied:
(Also, a great example of why I’ve been avoiding rebuilds.)

sternenseemann replied:
This is correct now (in the LLVM 15 removal commit).

alyssais commented on pkgs/development/compilers/llvm/common/clang/default.nix:
```diff
@@ -118,7 +103,7 @@ stdenv.mkDerivation (
     ]
-    ++ lib.optionals (lib.versionAtLeast release_version "17") [
+    ++ [
       (lib.cmakeBool "LLVM_INCLUDE_TESTS" false)
```
More rebuild avoidance. (Making sure it's noted.)

alyssais commented on pkgs/development/compilers/llvm/common/lldb/default.nix:
```diff
@@ -146,7 +130,7 @@ stdenv.mkDerivation (
     ]
-    ++ lib.optionals (lib.versionAtLeast release_version "17") [
+    ++ [
       (lib.cmakeFeature "CLANG_RESOURCE_DIR" "../../../../${lib.getLib libclang}")
```
And here

alyssais reviewed (commented):
Lovely :)

emilazy commented on pkgs/development/compilers/llvm/common/bintools.nix:
```diff
@@ -22,42 +22,38 @@ runCommand "llvm-binutils-${version}"
-      for prog in ${llvm}/bin/*; do
-        ln -sf $prog $out/bin/${targetPrefix}$(basename $prog)
-      done
+  (''
```
Will fix.

emilazy commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -457,37 +370,25 @@ let
-        // lib.optionalAttrs (
-          lib.versionAtLeast metadata.release_version "15" && stdenv.targetPlatform.isWasm
-        ) { nixSupport.cc-cflags = [ "-fno-exceptions" ]; };
+        # FIXME: This should be inside the `wrapCCWith` call.
```
That would cause rebuilds. While working on this I already occasionally accidentally read conditionals inverted and dropped them or made them unconditonal unintentionally. As my intent is only to simplify and refactor the existing derivation, I want to ensure I don’t cause any rebuilds with this PR, only removed packages.

emilazy commented on pkgs/development/compilers/llvm/common/libunwind/default.nix:
```diff
@@ -19,47 +19,35 @@
 stdenv.mkDerivation (
   finalAttrs:
   let
+    # TODO: Is this still relevant?
```
Will drop it, as many other derivations don’t have it.

emilazy commented on pkgs/development/compilers/llvm/common/lldb/default.nix:
```diff
@@ -218,7 +148,7 @@ stdenv.mkDerivation (
           return 1;
       fi
     '' # Something lua is built on older versions but this file doesn't exist.
```
Will drop.

emilazy commented on pkgs/development/compilers/mrustc/bootstrap.nix:
```diff
@@ -4,7 +4,7 @@
   mrustc,
   mrustc-minicargo,
-  llvm_12,
+  #llvm_12,
```
For cases like these where I hope the packages can be revived and don’t want to immediately remove the valuable knowledge encoded in them from the tree, I’ve erred on the side of preserving as much of the original source as possible without breaking eval. This, nominally, documents that LLVM 12 was previously being used for this. I did similar for the TensorFlow source build.

However I don’t feel strongly about this and can adjust as desired. I think you can make an argument for removing files like these entirely, but my hope is that they signal something useful to people looking into things like “can I bootstrap Rust with our `mrustc` package?”, and might get picked up before they rot too heavily.

emilazy commented on pkgs/development/compilers/llvm/common/lldb/default.nix:
```diff
@@ -72,7 +72,7 @@ stdenv.mkDerivation (
     ];
 
-    sourceRoot = lib.optional (lib.versionAtLeast release_version "13") "${finalAttrs.src.name}/lldb";
+    sourceRoot = [ "${finalAttrs.src.name}/lldb" ];
```
Yeah, rebuilds again. Do you want me to add a comment?

alyssais replied:
I think it would be nice to have comments to make the job of whoever ends up doing the staging cleanup easier.

emilazy commented on pkgs/development/compilers/llvm/common/lld/default.nix:
```diff
@@ -67,23 +61,16 @@ stdenv.mkDerivation (finalAttrs: {
-      '(''${LLVM_MAIN_SRC_DIR}/' '(../'
-  '';
+  # TODO: Remove on `staging`.
+  postPatch = "";
```
Yes, per earlier comment. Rebuilds obscure whether I messed anything up here.

emilazy commented on pkgs/development/compilers/llvm/update.sh:
```diff
@@ -20,36 +20,10 @@ sed -Ei \
-if [ "$VERSION_MAJOR" -ge "14" ]; then
-  readonly SOURCES=(
-    "llvm.monorepoSrc"
-  )
```
Look below: we are inlining it, because now `[ "$VERSION_MAJOR" -ge "14" ]` is always true. However, it’s true that some of these lines can be removed in the LLVM 12 drop, so I’ll squash those ones down.

These were already broken, by the way, because `monorepoSrc` is always used these days… but I am trying to be fairly mechanical here for ease of validation and review; can’t clean up everything weird about this package set in one go, but at least there’ll be less historical junk lying around.

alyssais commented on pkgs/development/compilers/llvm/common/lldb/default.nix:
```diff
@@ -149,7 +142,7 @@ stdenv.mkDerivation (
-    ++ lib.optionals (lib.versionAtLeast release_version "14") [
+    ++ [
       ninja
     ];
```
Yeah, I saw those, and that sort of thing is the reason I try in reviews of rebuild-avoiding PRs to get a commitment to cleaning them up on staging as soon as the PR makes its way there.  Admittedly the situation is a bit different here though, where a lot of them are being removed as well, and there's a legitimate reason to aim for 0 rebuilds, rather than just impatience.

emilazy replied:
Right. Do you want me to commit to that here? I feel like the ordering of some list concatenations is not too distasteful, and while I don’t love the empty lists and strings they’re a lot better than `${if … then … else null} = …;` contortions.

My feeling is that there are already a lot of things like this in the LLVM packaging, many with explicit comments about needing clean‐up, so it’s overdue for someone going over these in general. I can certainly add comments to these if you’d like. If you want me to clean up the specific things I’m introducing here on `staging` then I can try to do that, but I definitely don’t want to add cleaning up all the pre‐existing instances to my backlog; frankly this whole project ended up taking longer than expected and I’m running out of steam a little, so I would prefer to get this landed and then switch gears to other work I’ve been neglecting as soon as possible.

sternenseemann commented on pkgs/development/compilers/llvm/common/openmp/default.nix:
```diff
@@ -25,114 +25,91 @@
+    "dev"
+  ];
 
-    patchFlags = if lib.versionOlder release_version "14" then [ "-p2" ] else null;
```
Doesn't this removal cause a rebuild?

sternenseemann commented on pkgs/development/compilers/llvm/common/openmp/default.nix:
```diff
@@ -25,114 +25,91 @@
-    ++ lib.optionals (lib.versionAtLeast release_version "15") [
-      ninja
-    ]
-    ++ lib.optionals (lib.versionAtLeast release_version "14") [
```
This can't be folded into the previous conditional in the LLVM 13 removal, but needs to be its own, unconditional list.

emilazy replied:
Good catch, thank you. Will split this up between the two commits.

sternenseemann commented on pkgs/development/compilers/llvm/15/clang/purity.patch:
```diff
@@ -11,13 +11,12 @@ diff --git a/lib/Driver/ToolChains/Gnu.cpp b/lib/Driver/ToolChains/Gnu.cpp
 index fe3c0191bb..c6a482bece 100644
 --- a/lib/Driver/ToolChains/Gnu.cpp
 +++ b/lib/Driver/ToolChains/Gnu.cpp
```
Why does this patch for LLVM 15 change when removing LLVM 14?

emilazy replied:
More misunderstanding of the semantics of `getVersionFile` on my part. Will fix.

sternenseemann commented on pkgs/development/compilers/llvm/common/lldb/default.nix:
```diff
@@ -78,23 +78,7 @@ stdenv.mkDerivation (
-          hash = "sha256-u0zSejEjfrH3ZoMFm1j+NVv2t5AP9cE5yhsrdTS1dG4=";
-        })
-
-        # FIXME: do we need this after 15?
```
Do we…?

emilazy replied:
Apparently not, since we’ve gone this long without it? It was added in 99eeff0f69c527b1cf4b7eb49388955eaab9723f without explanation. I guess it was intended to resolve… an issue with Musl on PowerPC defining `pt_regs`?

I don’t know if the LLDB source would still run into this, but [this Musl commit](https://git.musl-libc.org/cgit/musl/commit/arch?h=v1.1.24&id=6ad514e4e278f0c3b18eb2db1d45638c9af1c07f) implies that it has not been an issue since 2019, regardless of LLVM version. In any case I think we would not accept a patch for an issue with PowerPC and Musl today without it being sent upstream to LLVM or Musl.

The i686 stuff in that patch, I don’t know. It’s not even related to the `<sys/procfs.h>` issue so I don’t know why it was added there in b4ee532f360090c3ed3277e7224763bd6429b412, but I think LLDB on i686 is also marginal enough that any relevant fixes should go upstream, if they’re even still needed. cc @rrbutani who added it.

In any case it would cause a rebuild to port it forward here, of course.

sternenseemann commented on pkgs/development/compilers/llvm/17/libclc/gnu-install-dirs.patch:
```diff
@@ -0,0 +1,10 @@
+-includedir=@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_INCLUDEDIR@
+-libexecdir=@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_DATADIR@/clc
++includedir=@CMAKE_INSTALL_FULL_INCLUDEDIR@
++libexecdir=@CMAKE_INSTALL_FULL_DATADIR@/clc
```
Was this patch accidentally deleted in the LLVM 15 removal commit and re-added in the LLVM 16 one?

Seems wrong that it's added and not moved from somewhere.

emilazy replied:
Rather, it looks like the commit should delete it from the LLVM 16 tree too. Not sure how that got lost.

sternenseemann commented on pkgs/development/compilers/llvm/common/libunwind/default.nix:
```diff
@@ -60,15 +60,6 @@ stdenv.mkDerivation (finalAttrs: {
   ]
   ++ devExtraCmakeFlags;
 
-  prePatch = lib.optionalString (lib.versionOlder release_version "18") ''
```
Doesn't the removal of `prePatch` and `postPatch` cause a rebuild?

emilazy replied:
Yes. Will fix.

sternenseemann reviewed (commented):
Nice to see some of the ugly conditionals disappear (e.g. from #347887 and #348568 🤐)!

- 54d55876a456180a837ca467b6f6124e199cab15 seems to prematurely remove some patches that belong in the LLVM 15 removal commit.

emilazy commented on pkgs/by-name/sp/spirv-llvm-translator/package.nix:
```diff
@@ -70,15 +50,6 @@ stdenv.mkDerivation {
     inherit (branch) rev hash;
   };
 
-  patches = lib.optionals (llvmMajor == "14") [
```
Rebuilds Mesa, even, though Mesa isn’t a big rebuild these days. But I’ll put it back with a TODO to stick to the aim of only having this PR drop packages, not rebuild any.

emilazy commented on pkgs/development/compilers/llvm/common/lldb/default.nix:
```diff
@@ -258,7 +258,7 @@ stdenv.mkDerivation (
     pname = "lldb-manpages";
 
     buildPhase = lib.optionalString (lib.versionOlder release_version "15") ''
-      make ${if (lib.versionOlder release_version "12") then "docs-man" else "docs-lldb-man"}
```
Right, good catch. I ran into a few of these which is what caused the split commit to be created, but apparently failed to move them all back.

emilazy commented on pkgs/development/compilers/fpc/default.nix:
```diff
@@ -111,5 +111,9 @@ stdenv.mkDerivation rec {
+    # See:
+    # * <https://gitlab.com/freepascal.org/fpc/source/-/issues/41045>
+    # * <https://gitlab.com/freepascal.org/fpc/source/-/merge_requests/887>
+    broken = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isx86_64;
```
I don’t think I had any reason to make that change, so I’ve just put it back to looking at Clang instead.

emilazy commented on pkgs/by-name/sh/shader-slang/package.nix:
```diff
@@ -73,8 +73,8 @@ stdenv.mkDerivation (finalAttrs: {
-    llvmPackages_14.llvm
-    llvmPackages_14.libclang
+    llvmPackages.llvm
+    llvmPackages.libclang
```
Done.

emilazy commented:
This now includes the temporary `meta.broken` for the LLVM backend of GHC, as discussed with Haskell maintainers in the relevant PRs, but should not change or regress any Haskell builds that are part of the release jobset. I have added `meta.broken` to `opencl-clang`/`intel-graphics-compiler` here as well, to check eval and as a fallback option given the uncertainty around https://github.com/NixOS/nixpkgs/pull/440272; my patch in https://github.com/NixOS/nixpkgs/pull/440534#pullrequestreview-3193917992 seems like the most viable way to keep the package working if we don’t proceed with the drop. Everything else has been fully dealt with already.

@athas I see you commented on https://github.com/jrprice/Oclgrind/issues/209; do you think there’s a realistic chance of it getting updated upstream to support recent LLVM any time soon (preferably LLVM 19 or higher, as there’s a decent chance we’ll drop 18 in 26.05)? If so, I’m happy to mark it as `broken` rather than dropping for now – I saw there had been no commits for three years, so I wasn’t too hopeful about the prospect of continued upstream maintenance. If not, an out‐of-tree package as described in https://github.com/jrprice/Oclgrind/issues/209#issuecomment-3263684989 seems like a good solution to me.

@wolfgangwalther I do not understand the listed rebuilds for `haskellPackages.GlomeVec` and `haskellPackages.spaceprobe`. Any insights?

@philiptaron @alyssais @sternenseemann I believe I have addressed all of your review comments that I haven’t explicitly responded to saying otherwise; would be great if you can double‐check my work and mark threads resolved as appropriate :)

emilazy commented:
Everything required here has now been merged into `master`, and this rebuilds only `tests.cc-wrapper.supported` and things downstream of the manual, as expected. I also added a release note. I’ll leave the button for someone else to push :)
